File: app/lib/core/google_auth.dart
import 'dart:async';

import 'package:flutter/foundation.dart' show kIsWeb;
import 'package:google_sign_in/google_sign_in.dart';
import 'package:google_sign_in_platform_interface/google_sign_in_platform_interface.dart'
    show
        AuthorizationRequestDetails,
        GoogleSignInPlatform,
        ServerAuthorizationTokensForScopesParameters;

import 'settings.dart';

/// Parent-side Google sign-in (SPEC 12: a child never signs in to anything).
///
/// One consent covers identity and `youtube.readonly`, which is what lets the
/// parent import the channels they already follow. What leaves the device is
/// the **server auth code**, never an access or refresh token: the gateway
/// exchanges the code and keeps the refresh token (docs/PROTOCOL.md).
///
/// Written against google_sign_in 7.2.0, whose API is not the 5.x/6.x one:
/// `GoogleSignIn.instance`, `initialize(...)`, `authenticate(...)`, and a
/// separate `authorizationClient` for scopes and server access.
///
/// **The web takes a different route to the same place.** GIS will not let an
/// app authenticate from its own button, and its identity button hands back
/// only an id_token — the YouTube scope then needs a second popup, and so a
/// second tap, for something a parent has already agreed to. So a browser skips
/// identity as a separate step and goes straight to the authorization-code
/// flow, which asks for the account and the scope in one window. The code that
/// comes back is exchanged for a refresh token exactly as a phone's is, and the
/// gateway reads who the parent is out of that exchange. One button, one popup,
/// on every platform.
class GoogleAuth {
  GoogleAuth({String? serverClientId})
    : serverClientId = serverClientId ?? BuildConfig.googleServerClientId;

  /// The one instance the app uses. `initialize` may only be called once per
  /// process, so the UI must not build a fresh [GoogleAuth] per button press.
  static final GoogleAuth shared = GoogleAuth();

  /// The **web** OAuth client id. Android needs it as `serverClientId` before
  /// the platform will issue a server auth code at all; in a browser the same
  /// id is the `clientId`, because there the app *is* the web client.
  final String serverClientId;

  /// PROTOCOL: the only scope we ask for. Read-only, the parent's own account.
  static const youtubeReadonlyScope =
      'https://www.googleapis.com/auth/youtube.readonly';

  /// Google's own reserved word for a code minted by a popup, which is the
  /// only flow a browser has. A phone sends nothing instead.
  static const popupRedirect = 'postmessage';

  /// What a browser's one popup asks for.
  ///
  /// The YouTube scope is the same one a phone asks for. The three identity
  /// scopes come with it because the browser has no separate sign-in step to
  /// get them from, and without an `id_token` the gateway cannot tell which
  /// Google account this is — it would mint a fresh household on every sign-in
  /// rather than returning the parent to their own.
  static const webScopes = ['openid', 'email', 'profile', youtubeReadonlyScope];

  /// False when the build carries no `HEYGILLI_GOOGLE_SERVER_CLIENT_ID`. The
  /// UI then hides the button instead of showing one that cannot work.
  bool get isConfigured => serverClientId.isNotEmpty;

  /// `initialize` must be called exactly once per process, so the future is
  /// cached and awaited rather than the call repeated.
  Future<void>? _initialized;

  Future<void> _initialize() {
    return _initialized ??= GoogleSignIn.instance.initialize(
      // In a browser the id identifies this app to Google directly; on a
      // phone it names the server the auth code is minted for.
      clientId: kIsWeb ? serverClientId : null,
      serverClientId: kIsWeb ? null : serverClientId,
    );
  }

  /// A browser, in one window: pick the account and grant the scope together.
  ///
  /// Goes to the platform directly because the high-level API has no way to
  /// ask for a server auth code without an already-authenticated account, and
  /// requiring one is exactly the second tap this avoids. With no `userId` the
  /// SDK prompts for the account as part of the flow, which is what makes it a
  /// single window.
  /// How long to wait on the popup before saying something.
  ///
  /// Long, because this is a person reading: the account chooser, then the
  /// unverified-app warning while the project is in review, then consent.
  /// Bounded, because a popup the browser blocked never returns at all, and
  /// waiting on it forever left the button spinning with nothing said — which
  /// is what it did, silently, for the whole of one screen recording.
  static const popupWait = Duration(minutes: 3);

  Future<GoogleAuthResult> _webSignIn() async {
    try {
      final tokens = await GoogleSignInPlatform.instance
          .serverAuthorizationTokensForScopes(
            const ServerAuthorizationTokensForScopesParameters(
              request: AuthorizationRequestDetails(
                scopes: webScopes,
                userId: null,
                email: null,
                promptIfUnauthorized: true,
              ),
            ),
          )
          .timeout(popupWait);
      // Null is the parent closing the window, which is a decision, not a fault.
      if (tokens == null || tokens.serverAuthCode.isEmpty) {
        return const GoogleAuthCancelled();
      }
      // Name and email are left to the gateway: they come out of the same
      // token exchange, so asking Google for them twice would be a second
      // round trip for something the next response already carries.
      return GoogleAuthSuccess(
        serverAuthCode: tokens.serverAuthCode,
        email: '',
        redirectUri: popupRedirect,
      );
    } on TimeoutException {
      return const GoogleAuthFailed(
        'Google never answered. If no window opened, your browser probably '
        'blocked the pop-up — allow pop-ups for this site and try again.',
      );
    }
  }

  /// The half of the flow after Google knows who the parent is: ask for the
  /// scope with offline access and get back the code the gateway exchanges.
  ///
  /// **Call this from a button press.** Where
  /// `authorizationRequiresUserInteraction()` is true — the web — anything
  /// else is a popup the browser will not open.
  Future<GoogleAuthResult> authorize(GoogleSignInAccount user) async {
    try {
      // Can legitimately return null when the platform has no server auth code
      // to give. PROTOCOL requires one, so that is a failure for us, not a
      // silent success.
      final server = await user.authorizationClient.authorizeServer(const [
        youtubeReadonlyScope,
      ]);
      if (server == null || server.serverAuthCode.isEmpty) {
        return const GoogleAuthScopeDenied();
      }
      return GoogleAuthSuccess(
        serverAuthCode: server.serverAuthCode,
        email: user.email,
        displayName: user.displayName ?? '',
        redirectUri: kIsWeb ? popupRedirect : '',
      );
    } catch (e) {
      return _asResult(e);
    }
  }

  /// Every failure the SDK can raise, as one of our own results. Nothing that
  /// reaches a screen is ever a raw platform exception.
  static GoogleAuthResult _asResult(Object e) {
    if (e is! GoogleSignInException) return GoogleAuthFailed('$e');
    return switch (e.code) {
      GoogleSignInExceptionCode.canceled ||
      GoogleSignInExceptionCode.interrupted => const GoogleAuthCancelled(),
      GoogleSignInExceptionCode.clientConfigurationError ||
      GoogleSignInExceptionCode.providerConfigurationError =>
        const GoogleAuthNotConfigured(
          'Google rejected this build\'s client id. On the web that is '
          'usually the page\'s address missing from the client\'s '
          'Authorized JavaScript origins.',
        ),
      GoogleSignInExceptionCode.uiUnavailable => const GoogleAuthUnavailable(
        'Google could not show its sign-in screen just now.',
      ),
      _ => GoogleAuthFailed(e.description ?? 'Google sign-in failed.'),
    };
  }

  /// Runs the whole flow and returns the server auth code.
  ///
  /// Must be called from a user interaction (a button press): the scope and
  /// server-authorization calls are allowed to show platform UI.
  ///
  /// Never throws a raw platform exception; every outcome is a typed result.
  Future<GoogleAuthResult> signIn() async {
    if (!isConfigured) return const GoogleAuthNotConfigured();
    try {
      await _initialize();

      if (!GoogleSignIn.instance.supportsAuthenticate()) return _webSignIn();

      // Authentication first; authorization is a separate step in 7.x.
      final user = await GoogleSignIn.instance.authenticate(
        scopeHint: const [youtubeReadonlyScope],
      );
      return authorize(user);
    } catch (e) {
      // MissingPluginException on a platform without the plugin, and anything
      // else the SDK throws: the caller only ever sees a result.
      return _asResult(e);
    }
  }

  /// Clears the local Google session. The gateway keeps its own link until the
  /// parent disconnects it there.
  Future<void> signOut() async {
    if (_initialized == null) return;
    try {
      await GoogleSignIn.instance.signOut();
    } catch (_) {
      // Signing out is best-effort; never block the UI on it.
    }
  }
}

/// Every outcome of [GoogleAuth.signIn], including the ones that are not bugs.
sealed class GoogleAuthResult {
  const GoogleAuthResult();
}

class GoogleAuthSuccess extends GoogleAuthResult {
  const GoogleAuthSuccess({
    required this.serverAuthCode,
    required this.email,
    this.displayName = '',
    this.redirectUri = '',
  });

  /// Sent to `POST /auth/google`. Single use, exchanged server-side.
  final String serverAuthCode;

  /// What Google minted the code against, travelling with the code because
  /// only the side that asked for it knows: empty from a phone, `postmessage`
  /// from a browser popup. The token exchange fails without a match.
  final String redirectUri;
  final String email;
  final String displayName;
}

/// The parent backed out. Not an error; say nothing.
class GoogleAuthCancelled extends GoogleAuthResult {
  const GoogleAuthCancelled();
}

/// No `HEYGILLI_GOOGLE_SERVER_CLIENT_ID` in this build, or the id does not
/// match the signing certificate.
class GoogleAuthNotConfigured extends GoogleAuthResult {
  const GoogleAuthNotConfigured([
    this.reason = 'Google sign-in is not set up in this build.',
  ]);
  final String reason;
}

/// Signed in, but YouTube access was not granted, so there is nothing to
/// import. The parent can still paste channel URLs.
class GoogleAuthScopeDenied extends GoogleAuthResult {
  const GoogleAuthScopeDenied();
}

/// Google Play services missing, no activity to show UI on, and similar.
class GoogleAuthUnavailable extends GoogleAuthResult {
  const GoogleAuthUnavailable(this.message);
  final String message;
}

class GoogleAuthFailed extends GoogleAuthResult {
  const GoogleAuthFailed(this.message);
  final String message;
}

File: agents/heygilli_agents/google_auth.py
"""Google sign-in for the parent and the YouTube read-only link.

PROTOCOL.md "Google sign-in and subscription import", SPEC §6.1 / §9.4 / §12.

The Android client shows the consent dialog and hands the gateway a **server
auth code**, never an access token. This module exchanges that code for a
refresh token using the *web* OAuth client credentials, caches the short-lived
access token beside it, and refreshes when it expires. The refresh token never
leaves the server and never goes back to the device.

One consent covers identity and nothing else: no sensitive scope, so no Google
verification review to keep alive. Only the parent's Google account is
involved: a child never
signs in to anything, and the parent's email is the single personal field
stored (SPEC §12).

Nothing in here logs a token, a refresh token, an auth code, or the payload
they travel in.

Environment (both required, both from the **web** OAuth client, not the
Android one):

    GOOGLE_CLIENT_ID
    GOOGLE_CLIENT_SECRET
"""
from __future__ import annotations

import base64
import binascii
import hashlib
import json
import logging
import os
import uuid
from dataclasses import dataclass
from datetime import UTC, datetime, timedelta

import httpx

from .schemas import GoogleLink, now_iso
from .store import Store

log = logging.getLogger(__name__)

TOKEN_URL = "https://oauth2.googleapis.com/token"
USERINFO_URL = "https://www.googleapis.com/oauth2/v3/userinfo"
HTTP_TIMEOUT_S = 20.0

NOT_CONFIGURED = (
    "Google sign-in is not configured on this server. Set GOOGLE_CLIENT_ID and "
    "GOOGLE_CLIENT_SECRET (the *web* OAuth client of the same Google Cloud project as the "
    "Android client) and restart the gateway. POST /auth/dev still works meanwhile."
)


class GoogleAuthError(Exception):
    """Something about the Google link is wrong, and the message says what."""


class GoogleNotConfigured(GoogleAuthError):
    """GOOGLE_CLIENT_ID / GOOGLE_CLIENT_SECRET are missing on the server."""


class GoogleNotLinked(GoogleAuthError):
    """This household has never signed in with Google (a normal state)."""


class GoogleNeedsRelink(GoogleAuthError):
    """The grant was revoked or expired: the parent has to sign in again."""


def credentials() -> tuple[str, str]:
    """The web OAuth client id and secret, read at call time so tests and
    `.env` reloads both work."""
    client_id = os.getenv("GOOGLE_CLIENT_ID", "").strip()
    client_secret = os.getenv("GOOGLE_CLIENT_SECRET", "").strip()
    if not client_id or not client_secret:
        raise GoogleNotConfigured(NOT_CONFIGURED)
    return client_id, client_secret


def configured() -> bool:
    return bool(os.getenv("GOOGLE_CLIENT_ID", "").strip() and os.getenv("GOOGLE_CLIENT_SECRET", "").strip())


@dataclass(frozen=True)
class GoogleTokens:
    """What the token endpoint returned. `refresh_token` is None on every
    exchange after the first one for the same account."""

    access_token: str = ""
    expires_in: int = 3600
    refresh_token: str | None = None
    id_token: str | None = None
    scope: str = ""


def expires_at(expires_in: int) -> str:
    return (datetime.now(UTC) + timedelta(seconds=max(0, int(expires_in)))).isoformat(timespec="seconds")


def _body(r: httpx.Response) -> dict:
    try:
        parsed = r.json()
    except ValueError:
        return {}
    return parsed if isinstance(parsed, dict) else {}


def _token_request(payload: dict[str, str], what: str) -> GoogleTokens:
    """One POST to Google's token endpoint.

    `payload` carries the auth code or the refresh token plus the client secret,
    so it is never logged, not even on failure.
    """
    try:
        with httpx.Client(timeout=HTTP_TIMEOUT_S) as client:
            r = client.post(TOKEN_URL, data=payload)
    except httpx.HTTPError as e:
        raise GoogleAuthError(f"could not reach Google to {what}: {type(e).__name__}") from e

    body = _body(r)
    if r.status_code >= 400:
        error = str(body.get("error") or f"http_{r.status_code}")
        detail = str(body.get("error_description") or "")
        log.warning("google token endpoint refused to %s: %s", what, error)
        if error == "invalid_grant":
            raise GoogleNeedsRelink(
                f"Google rejected the grant ({error}): it was revoked, already used or expired. "
                "The parent needs to sign in with Google again."
            )
        raise GoogleAuthError(f"Google could not {what}: {error} {detail}".strip())

    return GoogleTokens(
        access_token=str(body.get("access_token") or ""),
        expires_in=int(body.get("expires_in") or 3600),
        refresh_token=body.get("refresh_token") or None,
        id_token=body.get("id_token") or None,
        scope=str(body.get("scope") or ""),
    )


#: The only redirect a client may name. Google requires the token exchange to
#: match what the code was minted against, and there are exactly two cases:
#: a native consent has no redirect at all, and a browser popup uses Google's
#: own reserved word. Anything else is refused rather than forwarded, so this
#: endpoint cannot be talked into exchanging a code against a redirect of an
#: attacker's choosing.
POPUP_REDIRECT = "postmessage"
ALLOWED_REDIRECTS = frozenset({"", POPUP_REDIRECT})


def exchange_code(server_auth_code: str, redirect_uri: str = "") -> GoogleTokens:
    """Server auth code -> tokens. The credentials are always the **web** OAuth
    client's, whichever platform the code came from.

    `redirect_uri` is empty for a code minted by the Android or iOS client,
    which has no redirect and is refused if sent one. A browser's popup code
    must be exchanged against `postmessage`, or Google answers
    `invalid_request Missing parameter: redirect_uri` (PROTOCOL "Google sign-in
    and subscription import").
    """
    if redirect_uri not in ALLOWED_REDIRECTS:
        raise GoogleAuthError(
            f"Unsupported redirect_uri. Send {POPUP_REDIRECT!r} for a browser "
            "sign-in, or nothing at all for a phone."
        )
    client_id, client_secret = credentials()
    payload = {
        "grant_type": "authorization_code",
        "code": server_auth_code,
        "client_id": client_id,
        "client_secret": client_secret,
    }
    if redirect_uri:
        payload["redirect_uri"] = redirect_uri
    return _token_request(payload, "exchange the sign-in code")


def refresh_access_token(refresh_token: str) -> GoogleTokens:
    client_id, client_secret = credentials()
    return _token_request(
        {
            "grant_type": "refresh_token",
            "refresh_token": refresh_token,
            "client_id": client_id,
            "client_secret": client_secret,
        },
        "refresh the access token",
    )


def _id_token_claims(id_token: str) -> dict:
    """Read the claims out of Google's id_token without verifying the signature.

    That is safe here and only here: the token arrived over TLS as the direct
    response to our own token request, which is the one case OpenID Connect
    allows skipping verification.
    """
    try:
        payload = id_token.split(".")[1]
        claims = json.loads(base64.urlsafe_b64decode(payload + "=" * (-len(payload) % 4)))
    except (IndexError, ValueError, binascii.Error, UnicodeDecodeError):
        return {}
    return claims if isinstance(claims, dict) else {}


def _userinfo(access_token: str) -> dict:
    try:
        with httpx.Client(timeout=HTTP_TIMEOUT_S) as client:
            r = client.get(USERINFO_URL, headers={"Authorization": f"Bearer {access_token}"})
        return _body(r) if r.status_code == 200 else {}
    except httpx.HTTPError as e:
        log.warning("google userinfo failed: %s", type(e).__name__)
        return {}


def account_identity(tokens: GoogleTokens) -> tuple[str, str]:
    """`(google_sub, email)`. The id_token carries both when the client asked
    for the profile scopes; otherwise ask userinfo. Neither is fatal — the link
    works without an email to show."""
    claims = _id_token_claims(tokens.id_token) if tokens.id_token else {}
    sub, email = str(claims.get("sub") or ""), str(claims.get("email") or "")
    if not sub and tokens.access_token:
        info = _userinfo(tokens.access_token)
        sub, email = str(info.get("sub") or ""), str(info.get("email") or email)
    return sub, email


def household_id_for(sub: str, email: str) -> str:
    """Same Google account -> same household, every time. Matches the `hh_`
    shape `/auth/dev` produces."""
    seed = sub or email.strip().lower()
    if not seed:  # no identity claims at all: a fresh household is the honest answer
        return "hh_" + uuid.uuid4().hex[:10]
    return "hh_" + hashlib.sha1(f"google:{seed}".encode()).hexdigest()[:10]


def link_household(
    server_auth_code: str,
    store: Store,
    household_id: str | None = None,
    redirect_uri: str = "",
) -> tuple[str, GoogleLink]:
    """Exchange the code and persist the link. Returns `(household_id, link)`.

    Pass `household_id` to attach Google to a household the parent is already
    using (they started on `/auth/dev`); otherwise it is derived from the Google
    account so signing in twice lands in the same household.

    Google returns a refresh token only on the *first* consent for an account.
    A later exchange comes back without one, and the stored token is kept rather
    than overwritten with nothing.
    """
    tokens = exchange_code(server_auth_code, redirect_uri)
    sub, email = account_identity(tokens)
    hid = household_id or household_id_for(sub, email)
    stored = store.get_google_link(hid)

    refresh_token = tokens.refresh_token or (stored.refresh_token if stored else "")
    if not refresh_token:
        raise GoogleNeedsRelink(
            "Google returned no refresh token and none is stored for this household. "
            "The client must request offline access again (access_type=offline, prompt=consent)."
        )

    link = GoogleLink(
        email=email or (stored.email if stored else ""),
        google_sub=sub or (stored.google_sub if stored else ""),
        refresh_token=refresh_token,
        access_token=tokens.access_token,
        access_expires_at=expires_at(tokens.expires_in),
        linked_at=now_iso(),
    )
    store.put_google_link(hid, link)
    log.info("google link stored for household %s", hid)  # never the tokens themselves
    return hid, link


def google_access_token(household_id: str, store: Store) -> str:
    """A live access token for this household's YouTube link.

    Returns the cached one while it is still good, otherwise refreshes exactly
    once and caches the new one. A revoked or expired grant clears the link and
    raises `GoogleNeedsRelink` so the caller can say "sign in again" instead of
    turning it into a 500.
    """
    link = store.get_google_link(household_id)
    if link is None or not link.refresh_token:
        raise GoogleNotLinked("this household has not linked a Google account")
    if link.is_fresh():
        return link.access_token

    try:
        tokens = refresh_access_token(link.refresh_token)
    except GoogleNeedsRelink:
        store.clear_google_link(household_id)
        log.info("google grant rejected for household %s; link cleared", household_id)
        raise

    link.access_token = tokens.access_token
    link.access_expires_at = expires_at(tokens.expires_in)
    if tokens.refresh_token:  # Google does not rotate today, but honour it if it ever does
        link.refresh_token = tokens.refresh_token
    store.put_google_link(household_id, link)
    return link.access_token


def unlink(household_id: str, store: Store) -> None:
    """Forget the Google link (parent signs out, or data deletion, SPEC §12)."""
    store.clear_google_link(household_id)
